src/main/java/org/folio/auth/authtokenmodule/storage/RefreshTokenStore.java

```diff
@@ -0,0 +1,216 @@
+package org.folio.auth.authtokenmodule.storage;
+
+import org.folio.auth.authtokenmodule.tokens.RefreshToken;
+
+import java.time.Instant;
+import java.util.UUID;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.sqlclient.SqlConnection;
+import io.vertx.sqlclient.Tuple;
+
+/**
+ * The refresh token store provides persistence for refresh tokens. Refresh tokens need
+ * to be persisted so that: 1) we can ensure they are redeemed only once, and 2) we can
+ * take action (revoking all tokens for a user) when a given token is used more than once.
+ * When a refresh token is attempted to be redeemed more than once we consider it leaked.
+ */
+public class RefreshTokenStore extends TokenStore {
+  private static final Logger log = LogManager.getLogger(RefreshTokenStore.class);
+
+  private static final String REFRESH_TOKEN_SUFFIX = "refresh_tokens";
+
+  /**
+   * Constructs the refresh token store.
+   * @param vertx A reference to the current vertx object.
+   * @param tenant The tenant which is in scope for a given token. This can be obtained
+   * from the refresh token when the refresh token is created or arrives.
+   */
+  public RefreshTokenStore(Vertx vertx, String tenant) {
+    super(vertx, tenant);
+
+    // TODO implement this to clean up expired tokens from the store.
+    // TODO Consider using an okapi timer.
+    // vertx.setPeriodic(delay, handler)
+  }
+
+  /**
+   * Creates the table for this token store if it doesn't yet exist.
+   * @return A failed future should the save operation fail. Otherwise a succeeded future
+   * is returned, even if the table exists.
+   */
+  public Future<Void> createTableIfNotExists() {
+    // Refresh tokens have an owning user, but the token itself isn't persisted
+    // since it isn't used for anything. Just the id is enough.
+    String createTable = "CREATE TABLE IF NOT EXISTS " +
+        tableName(tenant, REFRESH_TOKEN_SUFFIX) +
+        "(id UUID PRIMARY key, user_id UUID NOT NULL, is_redeemed BOOLEAN NOT NULL, " +
+        "is_revoked BOOLEAN NOT NULL, expires_at INT8 NOT NULL)";
+
+    log.info("Creating {} tables", RefreshTokenStore.class.getName());
+
+    return pool.query(createTable).execute()
+      .mapEmpty();
+  }
+
+  public Future<Void> createIndexesIfNotExists() {
+    String createIndexExpiresAt = "CREATE INDEX IF NOT EXIST ON " +
+        tableName(tenant, REFRESH_TOKEN_SUFFIX) + "(expires_at)";
+    String createIndexUserId = "CREATE INDEX IF NOT EXIST ON " +
+        tableName(tenant, REFRESH_TOKEN_SUFFIX) + "(user_id)";
+
+    log.info("Creating {} tables", RefreshTokenStore.class.getName());
+
+    return pool.query(createIndexExpiresAt).execute()
+      .compose(x -> pool.query(createIndexUserId).execute())
+      .mapEmpty();
+  }
```

Comment:
createIndexesIfNotExists() can be merged into createTableIfNotExists(). It doesn't make sense to create the table without indexes.

Reply:
I broke this out into a separate method because I was having trouble having subsequent database operations complete after I created the indexes. In other words, I couldn't create the RT table, create its indexes, _and then_ create the API token table. Instead that third operation would silently fail. I didn't resolve this except to make the index creation the last thing done. Would love to know what might be going on there.